### PR TITLE
MG-182 - Secure token cookies

### LINKED ISSUE #182: Store secure cookies on the Browser

### PR DESCRIPTION
**Pull request by ianmuchyri**
# What type of PR is this?

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update
- [ ] Dependency Update

## What does this do?
Add an abstraction API redirect between login and viewing domains that encrypts the tokens that are produced during login and redirects to /domains with the cookies set which are encrypted
image

## Which issue(s) does this PR fix/relate to?

- Related Issue #
- Resolves #182 

## Have you included tests for your changes?

- [ ] Yes
- [x] No, and this is why: 

## Did you document any new/modified functionality?

- [ ] Yes
- [x] No, and this is why:

### CODE REVIEW

**Comment by ianmuchyri** [comment]
Dependent on this pr: #187

File: internal/postgres/postgres.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ var (
 // Config defines the options that are used when connecting to the PostgresSQL instance.
 type Config struct {
 	Host        string `env:"MG_UI_DB_HOST"           envDefault:"localhost"`
-	Port        string `env:"MG_UI_DB_PORT"           envDefault:"5432"`
+	Port        string `env:"MG_UI_DB_PORT"           envDefault:"6042"`
```
Comment (rodneyosodo) [inline internal/postgres/postgres.go:25]: Revert. Default Postgres port in `5432`

File: ui/api/transport.go [hunk 2]
```diff
@@ -60,16 +61,21 @@ const (
...
 	channelsItem            = "channels"
 	groupsItem              = "groups"
-	accessTokenKey          = "token"
+	accessTokenKey          = "access_token"
 	refreshTokenKey         = "refresh_token"
+	sessionDetailsKey       = "session_details"
 	channelKey              = "channel"
 	thingKey                = "thing"
+	loggedInKey             = "logged_in"
+	hashkey                 = "5jx4x2Qg9OUmzpP5dbveWQ"
```
Comment (rodneyosodo) [inline ui/api/transport.go:70]: These two should be declared are environment variables https://github.com/absmach/magistrala/blob/d95283d314720fc59035ebd3b254397ce948ab27/cmd/auth/main.go#L58

File: go.mod [hunk 3]
```diff
@@ -34,6 +34,7 @@ require (
 	github.com/gofrs/uuid v4.4.0+incompatible // indirect
 	github.com/golang/protobuf v1.5.3 // indirect
 	github.com/google/uuid v1.6.0 // indirect
+	github.com/gorilla/securecookie v1.1.2 // indirect
```
Comment (rodneyosodo) [inline go.mod:37]: Run go mod tidy

File: ui/service.go [hunk 4]
```diff
@@ -635,18 +653,49 @@ func (us *uiService) DomainLogin(login sdk.Login, refreshToken string) (sdk.Toke
...
+	}
+
+	data, err := json.Marshal(sessionDetails)
 	if err != nil {
-		return []byte{}, err
+		return "", err
 	}
 
-	return jsonData, nil
+	encodedString := base64.StdEncoding.EncodeToString(data)
```
Comment (rodneyosodo) [inline ui/service.go:696]: Base64 is not a secure way to encode information. Change this too

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/api/transport.go [hunk 5]
```diff
@@ -2424,87 +2505,110 @@ func tokenFromCookie(r *http.Request, cookie string) (string, error) {
...
+	}
 }
 
-func TokenMiddleware(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		tokenString, err := tokenFromCookie(r, accessTokenKey)
-		if err != nil {
-			if errors.Contains(err, http.ErrNoCookie) {
-				http.Redirect(w, r, "/token/refresh?referer_url="+url.QueryEscape(r.URL.String()), http.StatusSeeOther)
+func TokenMiddleware(s *securecookie.SecureCookie) func(http.Handler) http.Handler {
```
Comment (dborovcanin) [inline ui/api/transport.go:2544]: Can we do encryption and decryption of the cookies here or in a separate middleware (and replace request headers with decrypted cookies) so we make this change transparent from the service level?

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: cmd/ui/main.go [hunk 6]
```diff
@@ -39,6 +40,8 @@ type config struct {
 	InvitationsURL  string          `env:"MG_INVITATIONS_URL"     envDefault:"http://localhost:9020"`
 	MsgContentType  sdk.ContentType `env:"MG_UI_CONTENT_TYPE"     envDefault:"application/senml+json"`
 	TLSVerification bool            `env:"MG_UI_VERIFICATION_TLS" envDefault:"false"`
+	HashKey         string          `env:"MG_UI_HASH_KEY"         envDefault:"5jx4x2Qg9OUmzpP5dbveWQ"`
+	BlockKey        string          `env:"MG_UI_BLOCK_KEY"        envDefault:"/UtgZjr92jwRY6SPUndHXiyl9QY8qTUy"`
```
Comment (rodneyosodo) [inline cmd/ui/main.go:44]: Is this `/` needed?

```suggestion
	BlockKey        string          `env:"MG_UI_BLOCK_KEY"        envDefault:"UtgZjr92jwRY6SPUndHXiyl9QY8qTUy"`
```

File: docker/.env [hunk 7]
```diff
@@ -17,6 +17,8 @@ MG_UI_HOST_URL=http://localhost:9095
 MG_UI_VERIFICATION_TLS=false
 MG_UI_CONTENT_TYPE = application/senml+json
 MG_UI_INSTANCE_ID=
+MG_HASH_KEY=5jx4x2Qg9OUmzpP5dbveWQ
```
Comment (rodneyosodo) [inline docker/.env:20]:
```suggestion
MG_UI_HASH_KEY=5jx4x2Qg9OUmzpP5dbveWQ
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: docker/docker-compose.yml [hunk 8]
```diff
@@ -42,6 +42,8 @@ services:
       MG_UI_DB_SSL_CERT: ${MG_UI_DB_SSL_CERT}
       MG_UI_DB_SSL_KEY: ${MG_UI_DB_SSL_KEY}
       MG_UI_DB_SSL_ROOT_CERT: ${MG_UI_DB_SSL_ROOT_CERT}
+      MG_UI_HASH_KEY: ${MG_UI_HASH_KEY}
+      MG_UI_BLOCK_KEY: ${MG_UI_BLOCK_KEY}
```
[the remark this answers is not in the document]
Comment (rodneyosodo) [reply docker/docker-compose.yml:50]: Add the two variables to `ui/README.md`

File: ui/api/requests.go [hunk 9]
```diff
@@ -861,14 +880,15 @@ func (req domainLoginReq) validate() error {
...
-	sdk.Token
+	ui.Session
 	status string
 	page   uint64
 	limit  uint64
 }
 
 func (req listDomainsReq) validate() error {
 	if req.AccessToken == "" {
+		fmt.Println("here")
```
Comment (rodneyosodo) [inline ui/api/requests.go:891]: Remove

File: ui/service.go [hunk 10]
```diff
@@ -59,6 +58,9 @@ const (
 	channelsEndpoint        = "/channels"
 	groupsEndpoint          = "/groups"
 	bootstrapEndpoint       = "/bootstraps"
+	accessTokenKey          = "access_token"
+	refreshTokenKey         = "refresh_token"
+	sessionDetailsKey       = "session_details"
```
Comment (rodneyosodo) [inline ui/service.go:63]: are these used?

File: ui/service.go [hunk 11]
```diff
@@ -255,14 +258,14 @@ type Service interface {
 	// DomainLogin provides a user with an domain level access token and a refresh token.
 	DomainLogin(login sdk.Login, refreshToken string) (sdk.Token, error)
 	// Session retrieves the details of the user's session.
-	Session(token, session, domainID string) (string, error)
+	Session(s Session, loginStatus string) (Session, error)
```
Comment (rodneyosodo) [inline ui/service.go:261]: This `loginStatus` how is it different from `Session.LoginStatus`

File: ui/service.go [hunk 12]
```diff
@@ -255,14 +258,14 @@ type Service interface {
...
 
 	// CreateUsers creates new users.
 	CreateUsers(token string, users ...sdk.User) error
 	// ListUsers retrieves users owned/shared by a user.
-	ListUsers(token, status string, page, limit uint64) ([]byte, error)
+	ListUsers(status string, s Session, page, limit uint64) ([]byte, error)
```
Comment (rodneyosodo) [inline ui/service.go:266]: Let's keep `token` / `session` as first argument

File: ui/service.go [hunk 13]
```diff
@@ -472,6 +475,8 @@ func (us *uiService) Index(token string) (b []byte, err error) {
 		Status: enabled,
 	}
 
+	token := s.AccessToken
```
Comment (rodneyosodo) [inline ui/service.go:478]: Instead of creating a new variable just use `s.AccessToken`

File: ui/service.go [hunk 14]
```diff
@@ -682,34 +686,28 @@ func (us *uiService) Session(token, session, domainID string) (s string, err err
...
+	switch loginStatus {
 	case "user":
 		sessionDetails.LoginStatus = "user"
 	case "domain":
 		sessionDetails.LoginStatus = "domain"
-		domain, err := us.sdk.Domain(domainID, token)
+		domain, err := us.sdk.Domain(s.Domain.ID, s.AccessToken)
 		if err != nil {
-			return "", err
+			fmt.Println(err)
```
Comment (rodneyosodo) [inline ui/service.go:696]: remove debug statements

File: ui/service.go [hunk 15]
```diff
@@ -723,7 +721,8 @@ func (us *uiService) CreateUsers(token string, users ...sdk.User) error {
 	return nil
 }
 
-func (us *uiService) ListUsers(token, status string, page, limit uint64) ([]byte, error) {
+func (us *uiService) ListUsers(status string, s Session, page, limit uint64) ([]byte, error) {
+	token := s.AccessToken
```
Comment (rodneyosodo) [inline ui/service.go:725]: same

**Review by rodneyosodo — CHANGES_REQUESTED** [review]

File: ui/service.go [hunk 16]
```diff
@@ -381,9 +381,9 @@ type Service interface {
 	// UpdateBootstrapState updates bootstrap configuration state.
 	UpdateBootstrapState(token string, config sdk.BootstrapConfig) error
 	// ViewBootstrap retrieves a bootstrap config by thing id.
-	ViewBootstrap(token, thingID string) ([]byte, error)
+	ViewBootstrap(things Session, idID string) ([]byte, error)
```
Comment (rodneyosodo) [inline ui/service.go:384]:
```suggestion
	ViewBootstrap(s Session, id string) ([]byte, error)
```

File: ui/service.go [hunk 17]
```diff
@@ -531,10 +531,12 @@ func (us *uiService) Index(token string) (b []byte, err error) {
 		NavbarActive   string
 		CollapseActive string
 		Summary        dataSummary
+		SessionDetails Session
```
Comment (rodneyosodo) [inline ui/service.go:534]: Let's keep this aligned
```suggestion
		Session Session
```

**Review by rodneyosodo — CHANGES_REQUESTED** [review]
Also create a PR on magistrala since it changes https://github.com/absmach/magistrala/blob/0f05c10a3893c903222a65c974d0cd20da861103/docker/.env#L168

**Review by rodneyosodo — APPROVED** [review]

File: ui/service.go [hunk 18]
```diff
@@ -102,10 +101,11 @@ type Domain struct {
 	Permissions []string `json:"permissions"`
 }
 
-type SessionDetails struct {
+type Session struct {
 	User        User   `json:"user"`
 	Domain      Domain `json:"domain"`
 	LoginStatus string `json:"login_status"`
```
Comment (dborovcanin) [inline ui/service.go:107]: Make an enum.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

File: ui/service.go [hunk 19]
```diff
@@ -30,36 +30,40 @@ import (
...
+	membersActive                       = "members"
+	invitationsActive                   = "invitations"
+	domainInvitationsActive             = "domaininvitations"
+	usersEndpoint                       = "/users"
+	thingsEndpoint                      = "/things"
+	channelsEndpoint                    = "/channels"
+	groupsEndpoint                      = "/groups"
+	bootstrapEndpoint                   = "/bootstraps"
+	UserLoginStatus         LoginStatus = "user"
+	DomainLoginStatus       LoginStatus = "domain"
```
Comment (dborovcanin) [inline ui/service.go:62]: Extract to a separate `const` block.

**Review by dborovcanin — CHANGES_REQUESTED** [review]

**Review by dborovcanin — APPROVED** [review]

**Pull request merged**